### PR TITLE
#5 add config istio ecosystem chapter

### PR DESCRIPTION
**Pull request by Emily-Jiang**
Signed-off-by: Emily Jiang <emijiang@uk.ibm.com>

### CODE REVIEW

**Comment by Emily-Jiang** [comment]
please can someone review this @kenfinnigan @pilhuhn ?

**Comment by kenfinnigan** [comment]
Is it also worth mentioning that the same happens when the `microprofile-config.properties` file is mounted as a `ConfigMap`, saving the need to enter values directly?

File: spec/src/main/asciidoc/configuration.asciidoc [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+// limitations under the License.
+// Contributors:
+// Emily Jiang
+//
+
+= MicroProfile Config and Service Mesh
+
+MicroProfile Config provides a solution to externalise the configuration. The default config sources includes environment varialbles, system properties and microprofile-config.properties on the classpath. 
+
+The properties defined in Kubernetes config map automatically transforms to environment variables.
```
Comment (pilhuhn) [inline spec/src/main/asciidoc/configuration.asciidoc]: Not automatically. Only when you specify

```
- envFrom:
        - configMapRef:
            name: <name of the config map>
```

Comment (Emily-Jiang) [reply spec/src/main/asciidoc/configuration.asciidoc]: will update it.

**Comment by Emily-Jiang** [comment]
@kenfinnigan I will mention the mounting, but not for microprofile-config.properties. All config sources are already made themselves available to the microservices. If mounting the file again, we just get multiple properties. However, I am inspired by your comments and thought about using this to provide extra config sources without the need to repackaging the app.

**Comment by Emily-Jiang** [comment]
please re-review @kenfinnigan @pilhuhn

**Comment by kenfinnigan** [comment]
From my perspective I don't see this is the "best way" to incorporate MP Config into a Config Map, as it essentially converts properties in a file to environment variables, defeating the point of having a Config Map in the first place.

My preference, and I'm pretty sure this works, is to mount the properties file as a Config Map that can be read by the application as if it's coming from a ConfigSource.

To me that's more natural than somewhat "cheating" by exploding the file contents into a bunch of environment variables

**Comment by Emily-Jiang** [comment]
@kenfinnigan mapping a file to configmap is not the integration with MP Config. The whole integration is via env variables. Once configmap maps to env variables, the properties can be automatically be viewed by MP Config, so that it can be injected. Hence I put the whole paragraph about mapping without mentioning MP Config. Feel free to come with improved sentences, as I want to put the end to end understand:
file->configmap->env->mpconfig

**Comment by Emily-Jiang** [comment]
please review again @kenfinnigan @pilhuhn ! Thanks

File: spec/src/main/asciidoc/ecosystem.asciidoc [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+          envFrom:
+          - configMapRef:
+              name: example-config
+----
+
+Therefore, config properties specified in the config map are automatically injectable to the microservices via MicroProfile Config APIs.
+
+[source, java]
+----
+@ApplicationScoped
```
Comment (kenfinnigan) [inline spec/src/main/asciidoc/ecosystem.asciidoc:74]: I'm not in favor of having usages or sub spec details included in this document.

It's rife for becoming out of date too quickly. I would prefer to see links to other spec docs as appropriate.

File: spec/src/main/asciidoc/ecosystem.asciidoc [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        System.out.println(myProp1 + myProp2);
+    }
+}
+----
+The above example demonstrates that MicroProfile Config complements to service mesh via Kubernetes. 
+
+= MicroProfile Health Check and Service Mesh
+In Service Mesh architecture, each pod has a lifecycle. Service Mesh needs to know when to kill a pod and when to route requests to a pod. Therefore, it needs to know each pod's health status. A pod's health status is measured using Liveness and Readiness.
+
+=== Liveness
```
Comment (kenfinnigan) [inline spec/src/main/asciidoc/ecosystem.asciidoc:93]: Should these be "===" or "=="?

File: spec/src/main/asciidoc/ecosystem.asciidoc [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+=== Readiness
+
+Sometimes, microservices are temporarily unable to serve traffic. For example, an application might need to load large data or configuration files during startup. 
+
+=== MicroProfile Health Check
+
+MicroProfile Health Check denotes whether the microservice is live or ready. It exposes an endpoint `/health`. Invoking the endpoint returns either UP (healthy) or DOWN (unhealthy).
+ 
+[source, text]
+
```
Comment (kenfinnigan) [inline spec/src/main/asciidoc/ecosystem.asciidoc:105]: Again, specific examples are going to get out of date quickly, especially considering the changes being discussed around separate live and ready endpoints.

Linking to the particular spec document is preferred

File: spec/src/main/asciidoc/ecosystem.asciidoc [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+----
+
+=== Service Mesh usage of Liveness and Readiness
+
+Service Mesh e.g. Istio can utilise the readiness and liveness status from the underline component such as Kubernetes. 
+Kubernetes provides liveness probes or readiness probes to detect and remedy such situations. Kubernetes can check the pods frequently. 
+If the pod is not live, it will destroy the pod and start a new one. If the application is not ready, Kubernetes doesn’t want to kill the application, but not to send it requests either. 
+
+
+== The ecosystem
```
Comment (kenfinnigan) [inline spec/src/main/asciidoc/ecosystem.asciidoc:146]: Should this be "=="? It seems wrong given the previous headings

File: spec/src/main/asciidoc/ecosystem.asciidoc (hunk 3 above)
Comment (Emily-Jiang) [reply spec/src/main/asciidoc/ecosystem.asciidoc:93]: updated to ==

File: spec/src/main/asciidoc/ecosystem.asciidoc (hunk 2 above)
Comment (Emily-Jiang) [reply spec/src/main/asciidoc/ecosystem.asciidoc:74]: deleted

File: spec/src/main/asciidoc/ecosystem.asciidoc (hunk 5 above)
Comment (Emily-Jiang) [reply spec/src/main/asciidoc/ecosystem.asciidoc:146]: updated

File: spec/src/main/asciidoc/ecosystem.asciidoc (hunk 4 above)
Comment (Emily-Jiang) [reply spec/src/main/asciidoc/ecosystem.asciidoc:105]: updated

**Comment by Emily-Jiang** [comment]
I accidentally messed up this branch. I got all changes under #33 . Can you continue reviewing that PR?

**Comment by Emily-Jiang** [comment]
I messed up this branch. I have fixed all the comments under the PR 33

**Pull request closed without merging**